### ISSUE TITLE
[feature request] Defining gradient functions of new Ops using custom python functions

### ISSUE THREAD

**Issue by chenqifeng22** [tensorflow/tensorflow#5084]
As explained in https://www.tensorflow.org/versions/r0.11/how_tos/adding_an_op/index.html, the gradient functions for new Ops have to use other tensorflow Ops. This is a big limitation for me. I want to use other python functions to implement the gradient function.

Btw, this is a good [answer](http://stackoverflow.com/questions/39048984/tensorflow-how-to-write-op-with-gradient-in-python) to define new Ops purely in python, but it still has the limitation that the gradient function has to use other existing tensorflow Ops.

**Comment by asimshankar** [comment]
Hi @chenqifeng22 : Could you elaborate a bit as to why you need to define this in python only? If the gradient for your custom op cannot be implemented in terms of existing ops, then you can also define a custom op for the gradient and use that. For example, the gradient of the `Sqrt` operation is the [`SqrtGrad`](https://github.com/tensorflow/tensorflow/blob/c8a45a8e236776bed1d14fd71f3b6755bd63cc58/tensorflow/core/ops/math_ops.cc#L258) operation. Would something like that not work for you?

Calling back into python during graph execution for any operation (and consequently gradient computation) is going to create performance problems - greatly hampering the use of multiple threads, cores and devices.

**Comment by chenqifeng22** [comment]
Hi @asimshankar I am defining a new Op in which a Caffe model is called. Its return value can be computed by the forward inference of Caffe in tf.py_func following this [post](http://stackoverflow.com/questions/39048984/tensorflow-how-to-write-op-with-gradient-in-python). For the gradient function of the new Op, I want to use the backpropagation of Caffe to compute gradients. Otherwise, I need to take some efforts to convert the Caffe model to tensorflow.

Thanks for your suggestion about SqrtGrad. I will try to define a custom Op for the gradient.

**Comment by yaroslavvb** [comment]
@chenqifeng22 you can use py_func to define gradient function using Python
ops, as described here --
http://stackoverflow.com/questions/40060047/tensorflow-how-can-i-process-in-numpy-op-outputs-in-py-func-gradient/40065673#40065673

**Comment by chenqifeng22** [comment]
@yaroslavvb This looks like a great solution! Thanks!
